### PR TITLE
feat(RPC): check if the fee or gasPrice is enough

### PR DESCRIPTION
**Pull request by Flouse**
## Changes
1. check if the fee or fee_rate/gasPrice of the L2Transaction is enough
- check the fee of MetaContract::CreateAccount
- check the fee of SUDTTransfer
- check the gasPrice of Polyjuice L2TX
    gasPrice: Value of the gas for a transaction

return： if the fee is too low for acceptance, return  RpcError, such as:
`{"jsonrpc":"2.0","id":57,"error":{"code":-32602,"message":"Gas Price too low for acceptance, should more than polyjuice_minimum_gas_price(500 shannons)."}}',
  error: Error: Gas Price too low for acceptance, should more than polyjuice_minimum_gas_price(500 shannons).`

2. check the fee of WithdrawalRequest

## Config
The default `fee_rates` HashMap is empty, which means that `check_fee` will always be successful.

### Config Example in gw-config.toml
```toml
# Example
[mem_pool.fee_config]
# the minimal fee of meta contract:
# meta_contract_minimum_fee = fee_rate * meta_contract_fee_weight
meta_contract_fee_weight = 2
# the minimal fee of a native sudt transfer transaction:
# sudt_transfer_minimum_fee = fee_rate * meta_contract_fee_weight
sudt_transfer_fee_weight = 2
# the minimal fee of a withdrawal request:
# withdrawal_minimum_fee = fee_rate * withdraw_fee_weight
withdraw_fee_weight = 2

# HashMap<fee_sudt_id, fee_rate>
[mem_pool.fee_config.fee_rates]
# (CKB_SUDT_ID -> 500)
0x1 = 500
# (Other_sUDT_ID -> fee_rate)
0x2 = 600
```

### Note: 
1. `fee_rate` is known as the minimum gasPrice in Ethereum.
2. Polyjuice backend uses CKB_SUDT(default to 0x1) to pay fee by default.

### CODE REVIEW

**Comment by Flouse** [comment]
fix(gw-tools): add fee arg for withdraw SubCommand

and test withdrawal with low fee:
```bash
> ./bin/gw-tools withdraw \
  --amount 0 \
  --capacity 1000 \
  --config-path ./config.toml \
  --owner-ckb-address ckt1qyqf22qfzaer95xm5d2m5km0f6k288x9warqnhsf4m \
  --privkey-path ./deploy/private_key \
  --scripts-deployment-path ./deploy/scripts-deploy-result.json \
  --fee 0.00000999

[2021-11-17T08:05:12Z INFO  gw_tools::withdraw] withdrawal_request: WithdrawalRequest { raw: RawWithdrawalRequest { nonce: Uint32(0x0a000000), capacity: Uint64(0x00e8764817000000), amount: Uint128(0x00000000000000000000000000000000), sudt_script_hash: Byte32(0x0000000000000000000000000000000000000000000000000000000000000000), account_script_hash: Byte32(0xe26198694096599ad4c2e9610678bbef57757c019b9974b56e2589cf7a63cfcf), sell_amount: Uint128(0x00000000000000000000000000000000), sell_capacity: Uint64(0x00e40b5402000000), owner_lock_hash: Byte32(0x6f33df7e3fffb47794f9ca37643118ba79207ac92596a0555e9038828e547357), payment_lock_hash: Byte32(0x0000000000000000000000000000000000000000000000000000000000000000), fee: Fee { sudt_id: Uint32(0x01000000), amount: Uint128(0xe7030000000000000000000000000000) } }, signature: Bytes(0xf97f474ca000692711d345b3352c1f3c80bb75ab9b691e515cc3d848e2470dfc3af63356ec64726447240ce0ec3ff61ed407c91a10721e2d27c9f9074fc1acc001) }
[2021-11-17T08:05:12Z ERROR gw_tools] Withdrawal error: Invalid params: The fee is too low for acceptance, should more than withdrawal_base_fee.
```

**Comment by Flouse** [comment]
[Integration Test](https://github.com/nervosnetwork/godwoken-tests/actions/workflows/test.yml)
Integration test passed: 
https://github.com/nervosnetwork/godwoken-tests/runs/4278995713?check_suite_focus=true

File: crates/config/src/config.rs [hunk 1]
```diff
@@ -187,49 +187,86 @@ impl Default for OffChainValidatorConfig {
...
     withdraw_fee_weight: u8,
 }
+const DEFAULT_CKB_FEE_RATE: u64 = 500;
 impl FeeConfig {
     pub fn is_supported_sudt(&self, sudt_id: u32) -> bool {
-        self.supported_fee_sudt_ids.contains(&sudt_id)
+        self.fee_rates.contains_key(&sudt_id)
+    }
+    /// The defaut fee_sudt_id is 1, which is CKB
+    fn default_fee_rate(&self) -> u128 {
```
Comment (jjyr) [inline crates/config/src/config.rs:208]: Why we need a default fee rate?

File: crates/config/src/config.rs [hunk 2]
```diff
@@ -187,49 +187,86 @@ impl Default for OffChainValidatorConfig {
...
+        self.fee_rates
+            .get(&gw_common::builtins::CKB_SUDT_ACCOUNT_ID)
+            .unwrap_or(&DEFAULT_CKB_FEE_RATE)
+            .to_owned()
+            .into()
     }
     /// Get the minimal fee of meta contract
     pub fn meta_contract_base_fee(&self) -> u128 {
```
Comment (jjyr) [inline crates/config/src/config.rs:216]: Please pass sudt_id as argument to this method, we shouldn't using a `default` sudt to return fee

File: crates/config/src/config.rs [hunk 3]
```diff
@@ -187,49 +187,86 @@ impl Default for OffChainValidatorConfig {
...
     }
     /// Get the minimal gasPrice of Polyjuice contract
     pub fn polyjuice_base_gas_price(&self) -> u128 {
-        self.fee_rate as u128
+        self.default_fee_rate()
     }
 }
 impl Default for FeeConfig {
     fn default() -> Self {
+        let mut fee_rates = HashMap::new();
```
Comment (jjyr) [inline crates/config/src/config.rs:234]: Remove this Default, or we use the derive(Default)

**Review by jjyr — CHANGES_REQUESTED** [review]

File: crates/generator/src/backend_manage.rs [hunk 4]
```diff
@@ -16,6 +24,19 @@ pub struct BackendManage {
     backends: HashMap<H256, Backend>,
 }
 
+/// Get backend name from the validator path
+fn get_backend_name(validator_path: &str) -> Option<BackendName> {
+    if validator_path.contains("meta") {
```
Comment (jjyr) [inline crates/generator/src/backend_manage.rs:29]: We shouldn't use path to detect the script. What happend if the linux user named `meta`?

File: crates/rpc-server/src/registry.rs [hunk 5]
```diff
@@ -636,6 +672,20 @@ async fn submit_withdrawal_request(
     };
     let withdrawal_bytes = withdrawal_request.into_bytes();
     let withdrawal = packed::WithdrawalRequest::from_slice(&withdrawal_bytes)?;
+    log::debug!("{}", withdrawal);
```
Comment (jjyr) [inline crates/rpc-server/src/registry.rs:675]: Remove these logs before send a PR

**Review by jjyr — CHANGES_REQUESTED** [review]

**Comment by jjyr** [comment]
Please remove the empty file `polyjuice.mol`

File: crates/tools/src/main.rs [hunk 6]
```diff
@@ -452,8 +460,8 @@ fn run_cli() -> Result<()> {
                         .long("fee")
                         .takes_value(true)
                         .required(false)
-                        .default_value("0")
-                        .help("transfer fee"),
+                        .default_value("10000")
```
Comment (jjyr) [inline crates/tools/src/main.rs:463]:
```suggestion
                        .default_value("0")
```

File: crates/utils/src/fee.rs [hunk 7]
```diff
@@ -1,13 +1,95 @@
...
+                log::warn!(
+                    "Gas Price too low for acceptance, should more than polyjuice_base_gas_price({} shannons).",
+                    min_gas_price);
+                return Err(anyhow!("Gas Price too low for acceptance, should more than polyjuice_base_gas_price({} shannons).",
+                min_gas_price));
+            }
+            Ok(())
+        }
+        None => {
+            log::warn!("Found backend without name, new fee rules should be considered.");
```
Comment (jjyr) [inline crates/utils/src/fee.rs:88]: We should reject a unknown tx

**Review by jjyr — CHANGES_REQUESTED** [review]

File: crates/tools/src/main.rs (hunk 6 above)
Comment (jjyr) [reply crates/tools/src/main.rs:463]: Why 10000??? user should decide this value explicitly

**Comment by jjyr** [comment]
Please avoid to use log::info!("{}", ...); to output, we can't distinguish where the log come from in the production environment. Please add some hint such as log::info("[check tx fee] {}", ....)

File: crates/common/src/error.rs [hunk 8]
```diff
@@ -17,6 +17,8 @@ cfg_if::cfg_if! {
             Store,
             #[error("Invalid short address error")]
             InvalidShortAddress,
+            #[error("The sudt is not supported to used as fee")]
```
Comment (jjyr) [inline crates/common/src/error.rs:20]:
```suggestion
            #[error("Do not support pay fee with the current Simple UDT")]
```

File: crates/config/src/config.rs [hunk 9]
```diff
@@ -121,6 +122,7 @@ pub struct BlockProducerConfig {
 
 #[derive(Clone, Default, Debug, PartialEq, Serialize, Deserialize)]
 pub struct BackendConfig {
+    pub backend_type: String,
```
Comment (jjyr) [inline crates/config/src/config.rs:125]: The backend type should be a enumerator

File: crates/config/src/config.rs [hunk 10]
```diff
@@ -184,6 +186,103 @@ impl Default for OffChainValidatorConfig {
...
+    /// meta_contract_fee_weight = 2
+    /// sudt_transfer_fee_weight = 2
+    /// withdraw_fee_weight = 2
+    ///
+    /// [mem_pool.fee_config.fee_rates]
+    /// 0x1 = 500
+    /// ```
+    fn default() -> Self {
+        Self {
+            meta_contract_fee_weight: 2,
```
Comment (jjyr) [inline crates/config/src/config.rs:256]: Why the default weight is 2, do we have any reason?

File: crates/generator/src/backend_manage.rs [hunk 11]
```diff
@@ -16,6 +24,15 @@ pub struct BackendManage {
     backends: HashMap<H256, Backend>,
 }
 
+fn get_backend_type(validator_path: &str) -> Result<BackendType> {
+    match validator_path {
+        "meta" => Ok(BackendType::Meta),
```
Comment (jjyr) [inline crates/generator/src/backend_manage.rs:29]: We should implement the conversion on BackendType

```
impl From<String> for BackendType {
}
```

File: crates/generator/src/generator.rs [hunk 12]
```diff
@@ -561,6 +561,34 @@ impl Generator {
         }
     }
 
+    /// Get backend name by script_hash
+    pub fn get_backend_type<S: State + CodeStore>(
```
Comment (jjyr) [inline crates/generator/src/generator.rs:565]: Do we really need this function?, why can't just use `manage.get_backend().map(|backend| backend.type)`?

**Review by jjyr — CHANGES_REQUESTED** [review]

File: crates/config/src/config.rs (hunk 10 above)
Comment (Flouse) [reply crates/config/src/config.rs:256]: No, just a simple small default value.
Do you have any suggestion?

File: crates/generator/src/generator.rs (hunk 12 above)
Comment (Flouse) [reply crates/generator/src/generator.rs:565]: There is a `cloned` step in fn `load_backend`.
I think using `load_backend` directly is not efficient, so I created fn `get_backend_type`.

Comment (jjyr) [reply crates/generator/src/generator.rs:565]: This function increase the complexity, and do not actually improve the performance. If you really think the `cloned` affects performance, we should refactor the `get_backend` function, just copy the code to another function do not solve the problem.

I suggest we remove this function and use `get_backend` instead.

File: crates/config/src/config.rs (hunk 10 above)
Comment (jjyr) [reply crates/config/src/config.rs:256]: We can just use `#[derive(Default)]` on this structure.

File: crates/generator/src/generator.rs (hunk 12 above)
Comment (jjyr) [reply crates/generator/src/generator.rs:565]: This function increase the complexity, and do not actually improve the performance. If you really think the `cloned` affects performance, we should refactor the `get_backend` function, just copy the code to another function do not solve the problem.

I suggest we remove this function and use `get_backend` instead.

File: crates/rpc-server/src/registry.rs [hunk 13]
```diff
@@ -569,10 +571,30 @@ async fn execute_raw_l2transaction(
...
+) -> Result<BackendType, RpcError> {
+    let to_id = raw_l2tx.to_id().unpack();
+    let script_hash = state.get_script_hash(to_id)?;
+    let backend_type = generator
+        .load_backend(&state, &script_hash)
+        .map(|backend| backend.backend_type)
+        .ok_or(RpcError::Full {
+            code: INVALID_PARAM_ERR_CODE,
+            message: TransactionError::BackendNotFound { script_hash }.to_string(),
+            data: Some(Box::new("Please check to_id")),
```
Comment (jjyr) [inline crates/rpc-server/src/registry.rs:587]:
```suggestion
            data: Some(Box::new("Can't find backend for receiver account")),
```

File: crates/rpc-server/src/registry.rs [hunk 14]
```diff
@@ -581,35 +603,48 @@ async fn submit_l2transaction(
...
+            data: None,
+        });
+    }
+
+    // check tx fee or gasPrice of the l2tx
+    let raw_l2tx = tx.raw();
+    let backend_type = get_backend_type(tree, generator, &raw_l2tx)?;
+    gw_utils::fee::check_l2tx_fee(&mem_pool_config.fee_config, &raw_l2tx, backend_type).map_err(
+        |err| {
+            log::warn!("check_fee_ret err: {}", err);
```
Comment (jjyr) [inline crates/rpc-server/src/registry.rs:638]:
```suggestion
            log::debug!("check_fee_ret err: {}", err);
```

This is a normal behaviour, not a warning for Godwoken program

File: crates/rpc-server/src/registry.rs [hunk 15]
```diff
@@ -637,6 +673,25 @@ async fn submit_withdrawal_request(
...
 
+    // Check Fee
+    let fee_sudt_id = withdrawal.raw().fee().sudt_id().unpack();
+    let fee = withdrawal.raw().fee().amount().unpack();
+    let withdrawal_minimum_fee = mem_pool_config
+        .fee_config
+        .withdrawal_minimum_fee(fee_sudt_id)?;
+    if fee < withdrawal_minimum_fee {
+        let err_msg = format!(
+            "The fee is too low for acceptance, should more than withdrawal_minimum_fee({}).",
```
Comment (jjyr) [inline crates/rpc-server/src/registry.rs:684]:
```suggestion
            "Fee isn't enough, required fee for withdrawal: {}.",
```

File: crates/tools/src/main.rs (hunk 6 above)
Comment (jjyr) [inline crates/tools/src/main.rs:463]:
```suggestion
                        .default_value("0")
```

Comment (jjyr) [reply crates/tools/src/main.rs:463]: Please don't change a default value in a PR without any explanation.

File: crates/utils/src/fee.rs [hunk 16]
```diff
@@ -1,13 +1,89 @@
...
+    match backend_type {
+        BackendType::Meta => {
+            let meta_contract_args = MetaContractArgs::from_slice(raw_l2tx_args.as_ref())?;
+            let fee_struct = match meta_contract_args.to_enum() {
+                MetaContractArgsUnion::CreateAccount(args) => args.fee(),
+            };
+            let sudt_id = fee_struct.sudt_id().unpack();
+            let meta_contract_base_fee = fee_config.meta_contract_minimum_fee(sudt_id)?;
+            if fee_struct.amount().unpack() < meta_contract_base_fee {
+                let err_msg = format!("The fee is too low for acceptance, should more than meta_contract_minimum_fee({}).",
```
Comment (jjyr) [inline crates/utils/src/fee.rs:39]:
```suggestion
                let err_msg = format!("Fee isn't enough, required meta_contract fee: {}.",
```

File: crates/utils/src/fee.rs [hunk 17]
```diff
@@ -1,13 +1,89 @@
...
+                meta_contract_base_fee);
+                log::warn!("[check_l2tx_fee] {}", err_msg);
```
Comment (jjyr) [inline crates/utils/src/fee.rs:41]: Remove this, this shouldnt output a warning on Godwoken.

File: crates/utils/src/fee.rs [hunk 18]
```diff
@@ -1,13 +1,89 @@
...
+                return Err(anyhow!(err_msg));
+            }
+            Ok(())
+        }
+        BackendType::Sudt => {
+            let sudt_id = raw_l2tx.to_id().unpack();
+            if !fee_config.is_supported_sudt(sudt_id) {
+                return Err(anyhow!("This sudt is not supported to used as fee. Please use SudtERC20Proxy to transfer this sUDT instead."));
```
Comment (jjyr) [inline crates/utils/src/fee.rs:49]:
```suggestion
                return Err(anyhow!("Simple UDT is unsupported. Please use ERC20 transfer to instead."));
```

File: crates/utils/src/fee.rs [hunk 19]
```diff
@@ -1,13 +1,89 @@
...
+            }
+            let sudt_transfer_base_fee = fee_config.sudt_transfer_minimum_fee(sudt_id)?;
+            let sudt_args = SUDTArgs::from_slice(raw_l2tx_args.as_ref())?;
+            let fee_amount = match sudt_args.to_enum() {
+                SUDTArgsUnion::SUDTQuery(_) => 0u128,
+                SUDTArgsUnion::SUDTTransfer(args) => args.fee().unpack(),
+            };
+            if fee_amount < sudt_transfer_base_fee {
+                let err_msg = format!("The fee is too low for acceptance, should more than sudt_transfer_minimum_fee({}).",
```
Comment (jjyr) [inline crates/utils/src/fee.rs:58]:
```suggestion
                let err_msg = format!("Fee isn't enough, required sudt_transfer fee: {}.",
```

File: crates/utils/src/fee.rs [hunk 20]
```diff
@@ -1,13 +1,89 @@
...
+                sudt_transfer_base_fee);
+                log::warn!("[check_l2tx_fee] {}", err_msg);
```
Comment (jjyr) [inline crates/utils/src/fee.rs:60]: Remove this

File: crates/utils/src/fee.rs [hunk 21]
```diff
@@ -1,13 +1,89 @@
...
+            }
+            // Note: Polyjuice use CKB_SUDT to pay fee by default
+            let poly_args = raw_l2tx_args.as_ref();
+            let gas_price = u128::from_le_bytes(poly_args[16..32].try_into()?);
+            let min_gas_price =
+                fee_config.polyjuice_minimum_gas_price(gw_common::builtins::CKB_SUDT_ACCOUNT_ID)?;
+            if gas_price < min_gas_price {
+                let err_msg = format!("Gas Price too low for acceptance, should more than polyjuice_minimum_gas_price({} shannons).",
+                min_gas_price);
+                log::warn!("[check_l2tx_fee] {}", err_msg);
```
Comment (jjyr) [inline crates/utils/src/fee.rs:79]: Remove this

File: crates/utils/src/fee.rs [hunk 22]
```diff
@@ -1,13 +1,89 @@
...
+            if raw_l2tx_args.len() < (8 + 8 + 16 + 16 + 4) {
+                return Err(anyhow!("invalid PolyjuiceArgs"));
+            }
+            // Note: Polyjuice use CKB_SUDT to pay fee by default
+            let poly_args = raw_l2tx_args.as_ref();
+            let gas_price = u128::from_le_bytes(poly_args[16..32].try_into()?);
+            let min_gas_price =
+                fee_config.polyjuice_minimum_gas_price(gw_common::builtins::CKB_SUDT_ACCOUNT_ID)?;
+            if gas_price < min_gas_price {
+                let err_msg = format!("Gas Price too low for acceptance, should more than polyjuice_minimum_gas_price({} shannons).",
```
Comment (jjyr) [inline crates/utils/src/fee.rs:77]:
```suggestion
                let err_msg = format!("Gas Price is too low, required gas_price {}.",
```

**Review by jjyr — CHANGES_REQUESTED** [review]

**Review by jjyr — APPROVED** [review]

**Review by zeroqn — APPROVED** [review]

**Review by jjyr — APPROVED** [review]

**Pull request merged**